### ISSUE TITLE
[future] disable support for unencrypted connections to non-local IPs once Let's Encrypt is ready

### ISSUE THREAD

**Issue by lorenzhs** [glowing-bear/glowing-bear#533]
Problem: Unencrypted connections are unacceptable outside your local network, but self-signed certificates cause lots of issues on several platforms (especially mobile ones, but Desktop Chrome is also annoying).

Idea: Rename the "use encryption" checkbox to "Force unencrypted connection" and only allow it to be checked when the weechat host is an IP in the local network (i.e., a private IP range). This will have the following effect:

- IP in local network: :heavy_check_mark: 
- hostname that resolves to IP in local :heavy_multiplication_x: (see below)
- public IP: not supported unless you can get a certificate for it (possible, but rarely used). anyway, this is very rare.
- hostname that resolves to public IP: :heavy_multiplication_x: 

The idea is that if it's a public IP, you *need* encryption. Let's Encrypt will make that very very easy. We can even make a WeeChat script that sets everything up for you. Also, every server has a hostname (even if it's just <IP>.hosterfoobar.com or with one of those free subdomain providers)

If you use hostnames in your local network, then you're probably good enough with computers to also issue a certificate from your own private CA, or make your hostnames subdomains of a domain you own and can get a real cert.

I want to get rid of unencrypted connections, they're dangerous.

**Comment by arnottcr** [comment]
two requests:
0) allow uncrypted connections to the hostname "localhost"; this is a de facto standard for 127.0.0.1 and I imagine not everyone wants to type 127.0.0.1 every time.

1) allow uncrypted connections to tor hidden service hostnames⁰. as tor implements end to end crypto underneath the HTTP level, many HS operators elect to not use double crypto [tor + tls] and thus the sites look plaintext to the browser¹.

⁰: the following regex should catch all tor hidden service addresses: ^[a-z2-7]{16}\.onion$
¹: the connections it technically encrypted (tor client)-to-(tor client) not end-to-end, but if your tor client and the hidden service's tor client both run on their respective localhosts, then it is effectively end to end. however, if I run a tor client and you connect to it, note that this is NOT default behavior and requires a custom configuration, I can see all your traffic. it is for this reason that the browser thinks that HSs are unencrypted, because all the traffic is pre-decrypted.

**Comment by lorenzhs** [comment]
0) `localhost` should obviously be okay. If someone goes to the effort of remapping localhost to something else then we can't protect them any more anyway.
1) I'm not familiar enough with Tor routing to judge that, but I guess it makes sense. We might have to come back to you for testing it though, you could just be the only user ;)

**Comment by arnottcr** [comment]
That is great; I am definitely willing to help out with testing if required.

If you would like more information on tor hidden service addresses look at the specifications linked below, under section 1.5 [currently line 526]: https://spec.torproject.org/rend-spec

Not sure of your user base, but a similar issue may also come up with i2p users.

**Comment by lorenzhs** [comment]
bump

**Comment by arnottcr** [comment]
before you implement this, make sure that it is ipv6 compliant. you have not listed what public and private ip_s are, but make sure that they include at least local addresses [fc00::/7, fe80::/10, and ::1/128]

note regexes for ipv6 are hard, and it looks like glowing bear currently does not have great support for ipv6. I was unable to get it setup, and it kept throwing invalid links for tls help, but I can spin off this into another issue if I investigate it further.

**Comment by torhve** [comment]
Before you do, make sure you are actually testing gb and not just browser implementations. We just send the string to the browser, so if the browser supports it, gb supports it.

And normal IPv6 works just fine here, it's my preferred way to my own gb instance.

**Comment by arnottcr** [comment]
yep, it appears that I was wrong about the ipv6 stuff: everything looks good now.
